File: src/main/java/at/searles/parsing/Recognizer.java
package at.searles.parsing;

import at.searles.lexer.Tokenizer;
import at.searles.parsing.annotation.AnnotationRecognizer;
import at.searles.parsing.combinators.*;
import at.searles.parsing.printing.ConcreteSyntaxTree;
import at.searles.parsing.tokens.TokenRecognizer;
import at.searles.regex.CharSet;
import org.jetbrains.annotations.NotNull;

public interface Recognizer extends Recognizable {

    static Recognizer fromString(String string, Tokenizer tokenizer, boolean exclusive) {
        return new TokenRecognizer(string, tokenizer, exclusive);
    }

    /**
     * Creates a recognizer that will detect EOF (which is -1 returned from
     * the stream).
     */
    static Recognizer eof(Tokenizer tokenizer) {
        return new TokenRecognizer(tokenizer.token(CharSet.chars(-1)), false);
    }

    @NotNull
    ConcreteSyntaxTree print();

    default Recognizer then(Recognizer recognizer) {
        return then(recognizer, false);
    }

    default Recognizer then(Recognizer recognizer, boolean allowParserBacktrack) {
        return new RecognizerThenRecognizer(this, recognizer, allowParserBacktrack);
    }

    default <T> Parser<T> then(Parser<T> parser) {
        return then(parser, false);
    }

    default <T> Parser<T> then(Parser<T> parser, boolean allowParserBacktrack) {
        // corresponds to prefix.
        return new RecognizerThenParser<>(this, parser, allowParserBacktrack);
    }

    default <T, U> Reducer<T, U> then(Reducer<T, U> parser) {
        return then(parser, false);
    }

    // corresponds to prefix.
    default <T, U> Reducer<T, U> then(Reducer<T, U> parser, boolean allowParserBacktrack) {
        return new RecognizerThenReducer<>(this, parser, allowParserBacktrack);
    }

    default Recognizer or(Recognizer recognizer) {
        return new RecognizerOrRecognizer(this, recognizer);
    }

    default Recognizer rep() { //Caution in printer!
        return new RecognizerRep(this);
    }

    default Recognizer plus() { //Caution in printer!
        return this.then(rep());
    }

    default Recognizer opt() {
        return opt(false);
    }

    default Recognizer opt(boolean alwaysPrint) {
        return new RecognizerOpt(this, alwaysPrint);
    }

    /**
     * Creates a reducer for a possibly empty csv-alike structure
     */
    default <T> Reducer<T, T> join(Reducer<T, T> reducer) {
        return new ReducerJoin<>(this, reducer);
    }

    default <T> Reducer<T, T> joinPlus(Reducer<T, T> reducer) {
        return new ReducerJoinPlus<>(this, reducer);
    }

    default <A> Recognizer annotate(A category) {
        return new AnnotationRecognizer<>(category, this);
    }
}

File: src/main/java/at/searles/parsing/Recognizer.kt
package at.searles.parsing

import at.searles.lexer.Tokenizer
import at.searles.parsing.combinators.*
import at.searles.parsing.combinators.ext.ReducerJoin
import at.searles.parsing.combinators.ext.ReducerJoinPlus
import at.searles.parsing.printing.ConcreteSyntaxTree
import at.searles.parsing.ref.RefRecognizer
import at.searles.parsing.tokens.TokenRecognizer
import at.searles.parsingtools.common.Init
import at.searles.regexp.CharSet
import at.searles.regexp.Text

interface Recognizer: CanRecognize {
    override fun recognize(stream: ParserStream): Boolean

    fun print(): ConcreteSyntaxTree

    operator fun plus(right: Recognizer): Recognizer {
        return RecognizerThenRecognizer(this, right)
    }

    operator fun <T> plus(parser: Parser<T>): Parser<T> {
        // corresponds to prefix.
        return RecognizerThenParser(this, parser)
    }

    // corresponds to prefix.
    operator fun <T, U> plus(right: Reducer<T, U>): Reducer<T, U> {
        return this.toReducer<T>() + right
    }

    infix fun or(other: Recognizer): Recognizer {
        return RecognizerOrRecognizer(this, other)
    }

    infix fun <T> or(other: Reducer<T, T>): Reducer<T, T> {
        return this + Mapping.create<T, T> { it } or other
    }

    fun rep(): Recognizer {
        return RecognizerRep(this)
    }

    fun rep1(): Recognizer {
        return this + rep()
    }

    fun opt(): Recognizer {
        return RecognizerOpt(this)
    }

    fun flag(): Parser<Boolean> {
        return this + Init(true) or Init(false)
    }

    fun optAlwaysPrint(): Recognizer {
        return RecognizerOpt(this, alwaysPrint = true)
    }

    fun <T> init(value: T): Parser<T> {
        return this + Init(value)
    }

    /**
     * Creates a reducer for a possibly empty csv-alike structure
     */
    fun <T> join(reducer: Reducer<T, T>): Reducer<T, T> {
        return ReducerJoin(reducer, this)
    }

    fun <T> join1(reducer: Reducer<T, T>): Reducer<T, T> {
        return ReducerJoinPlus(reducer, this)
    }

    fun ref(label: String): Recognizer {
        return RefRecognizer(label).apply {
            this.ref = this@Recognizer
        }
    }

    fun <T> toReducer(): Reducer<T, T> {
        return RecognizerToReducer(this)
    }

    companion object {
        fun fromString(string: String, tokenizer: Tokenizer): Recognizer {
            val tokenId = tokenizer.add(Text(string))
            return fromToken(tokenId, string, tokenizer)
        }

        fun fromToken(tokenId: Int, tokenString: String, tokenizer: Tokenizer): Recognizer {
            return TokenRecognizer(tokenId, tokenizer, tokenString)
        }

        /**
         * Creates a recognizer that will detect EOF (which is -1 returned from
         * the stream).
         */
        fun eof(tokenizer: Tokenizer): Recognizer {
            val eofId = tokenizer.add(CharSet.chars(-1))
            return fromToken(eofId, "<eof>", tokenizer)
        }
    }
}